src/vs/workbench/contrib/files/browser/views/explorerView.ts

```diff
@@ -684,7 +684,27 @@ export class ExplorerView extends ViewPane {
 	}
 
 	collapseAll(): void {
-		this.tree.collapseAll();
+		const collapseWorkspaceFoldersFirst = this.configurationService.getValue<IFilesConfiguration>().explorer.collapseWorkspaceFoldersFirst;
+		if (!collapseWorkspaceFoldersFirst) {
+			this.tree.collapseAll();
+			return;
+		}
+
+		const treeInput = this.tree.getInput();
+		if (Array.isArray(treeInput)) {
+			let hasExpandedItemInsideFolder = treeInput.some(folder =>
```

Comment:
Minor polish: Can we maybe do this a bit simpler? 
* Instead of doing `Array.from`, we can just do `folder.children.forEach` I believe.
* `hasExpandedItemInsideFolder` should be a `const`
* I do not like the double `tree.colapseAll` call. Can we just have some `return` statement before if we did something, so we just have one call to the `tree.collapseAll` at the end

Reply:
I like using `some` because it will terminate early and also gets the intent across more directly in my view than forEach + modifying a local. 

I removed the double `collapseAll` call though.

Reply:
Makes perfect sense. Feel free to merge in.
Thanks for this great pt.

Reply:
Actually I changed the `some` too